### PR TITLE
Allow object path for update and delete

### PR DESCRIPTION
**Pull request by Iblis**
Draft for #248

Not done so far:

- [x] Allow using fullpath for other query types (only DataObject so far)
- [x] Object-Path in GraphlQL Schema should probably be called 'fullpath' instead of 'path'

### CODE REVIEW

**Comment by Iblis** [comment]
Sorry, this also contains my changes from #252 right now

File: src/GraphQL/DataObjectType/ElementDescriptorInputType.php [hunk 1]
```diff
@@ -47,6 +47,7 @@ public function build(&$config)
         $config['fields'] = [
             'type' => Type::string(),
             'id' => Type::int(),
+            'path' => Type::string()
```
Comment (weisswurstkanone) [inline src/GraphQL/DataObjectType/ElementDescriptorInputType.php:50]: please call it fullpath because fullpath = path + key

**Comment by weisswurstkanone** [comment]
can you clean it up?

File: src/GraphQL/DataObjectInputProcessor/ManyToManyObjectRelation.php [hunk 2]
```diff
@@ -42,8 +42,13 @@ public function process(Concrete $object, $newValue, $args, $context, ResolveInf
                     if (isset($newValueItemValue["type"]) && $newValueItemValue["type"] !== 'object') {
                         throw new ClientSafeException("expected object type");
                     }
-
-                    $element = \Pimcore\Model\Element\Service::getElementById('object', $newValueItemValue["id"]);
+                    
+                    if (isset($newValueItemValue["id"])) {
```
Comment (weisswurstkanone) [inline src/GraphQL/DataObjectInputProcessor/ManyToManyObjectRelation.php:46]: maybe a candidate for a trait

**Comment by weisswurstkanone** [comment]
Hello @Iblis, I like the idea!

However, for a complete solution we would have to cover all element types (asset, document, object) for mutations&queries. So a lot to do 😄 

Please go ahead 👍 

Thanks for your efforts!

File: src/GraphQL/Mutation/MutationType.php [hunk 3]
```diff
@@ -324,14 +327,21 @@ public function buildDataObjectMutations(&$config = [], $context = [])
                 $deleteField = [
                     'type' => $deleteResultType,
                     'args' => [
-                        'id' => ['type' => Type::nonNull(Type::int())],
-                    ], 'resolve' => static function ($value, $args, $context, ResolveInfo $info) use ($entity, $modelFactory, $me) {
+                        'id' => ['type' => Type::int()],
+                        'fullpath' => ['type' => Type::string()],
```
Comment (SimonBader) [inline src/GraphQL/Mutation/MutationType.php:331]: @weisswurstkanone is it good practice to define all anonymous function as static and pass the $this pointer as $me? As $this can be accessed within the anonymous function "for free" by just not mark it as static, right?

File: src/GraphQL/DataObjectInputProcessor/ManyToManyObjectRelation.php (hunk 2 above)
Comment (SimonBader) [reply src/GraphQL/DataObjectInputProcessor/ManyToManyObjectRelation.php]: I added the trait 'ElementIdentificationTrait' which gets the element from Service::getElementById() or Service::getElementByPath()). Both functions do handle objects, assets and documents.

File: src/GraphQL/Mutation/MutationType.php [hunk 4]
```diff
@@ -324,14 +327,21 @@ public function buildDataObjectMutations(&$config = [], $context = [])
...
+                    ], 'resolve' => static function ($value, $args, $context, ResolveInfo $info) use ($me) {
                         try {
```
Comment (SimonBader) [inline src/GraphQL/Mutation/MutationType.php:333]: @weisswurstkanone is it good practice to define all anonymous function as static and pass the $this pointer as $me? As $this can be accessed within the anonymous function "for free" by just not mark it as static, right?

File: src/GraphQL/Mutation/MutationType.php [hunk 5]
```diff
@@ -860,24 +871,18 @@ public function buildDeleteElementMutation(&$config, $context, $type)
...
-                        $element = null;
-
-                        if ($type == "asset") {
-                            $element = Asset::getById($id);
-                        } else if ($type == "document") {
-                            $element = Document::getById($id);
-                        } else if ($type == "object") {
-                            $element = DataObject::getById($id);
-                        }
+                        $element = $me->getElementByTypeAndIdOrPath($args, $type);
```
Comment (SimonBader) [inline src/GraphQL/Mutation/MutationType.php:886]: @weisswurstkanone I replaced all the different handling of asset, document and object by the new trait 'ElementIdentificationTrait' which gets the elements from the "standard Pimcore functions" Service::getElementById and Service::getElementByPath. 
That way we do not have to know implementation details from Pimcore and the Service will manage it for us.

Comment (weisswurstkanone) [reply src/GraphQL/Mutation/MutationType.php:886]: excellent. please let me know when it is ready to be reviewed

File: src/GraphQL/Resolver/QueryType.php [hunk 6]
```diff
@@ -204,7 +198,7 @@ public function resolveAssetGetter($value = null, $args = [], $context, ResolveI
             $this->getGraphQlService()->getLocaleService()->setLocale($args['defaultLanguage']);
         }
 
-        $assetElement = Asset::getById($args['id']);
+        $assetElement = $this->getElementByTypeAndIdOrPath($args, 'asset');
         if (!$assetElement) {
             return null;
         }
```
Comment (SimonBader) [inline src/GraphQL/Resolver/QueryType.php:204]: @weisswurstkanone Document query (resolveDocumentGetter) does already provide fullpath additional to id. In the Document query implementation the fullpath is named 'path' though, which would make it inconsistent to our naming.
So what shall we do:
- rename 'fullpath' back to 'path' in this PR
- rename the Document query 'path' to 'fullpath' which would break backward-compatibility

**Review by SimonBader — COMMENTED** [review]
@weisswurstkanone a first draft of the PR is ready to be reviewed. 
We do have a couple of questions though which would have to be answered first before we can finish up the entire functionality.

File: src/GraphQL/Mutation/MutationType.php (hunk 5 above)
Comment (SimonBader) [reply src/GraphQL/Mutation/MutationType.php:886]: A first draft of the PR is ready to be reviewed.
We do have a couple of questions though which would have to be answered first before we can finish up the entire functionality.

File: src/GraphQL/Resolver/QueryType.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (weisswurstkanone) [reply src/GraphQL/Resolver/QueryType.php:204]: mhm, I guess there are no alternatives because we want to keep pimcore 5.7 as minimum requirement.

Comment (weisswurstkanone) [reply src/GraphQL/Resolver/QueryType.php:204]: I would support both versions for now and mark the other one as deprecated (removed in 1.0). Similar to 
https://github.com/pimcore/data-hub/issues/266

**Comment by weisswurstkanone** [comment]
@SimonBader , could you please resolve the conflicts ? Thx!

File: src/GraphQL/DataObjectType/PimcoreObjectType.php [hunk 7]
```diff
@@ -87,6 +87,10 @@ public function build($context = [])
             [
                 'type' => Type::id(),
             ],
+            'fullpath' => [
```
Comment (weisswurstkanone) [inline src/GraphQL/DataObjectType/PimcoreObjectType.php:90]: I don't think that we should expose this by default. You can still use the configurator to add it.

File: tests/GraphQL/Traits/ElementIdentificationTraitTest.php [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+}
+
+class ElementIdentificationTraitTest extends TestCase
+{
+    const TRAIT_TO_TEST = '\Pimcore\Bundle\DataHubBundle\GraphQL\Traits\ElementIdentificationTrait';
+    const TEST_TYPE = 'object';
+
+    public function testThrowingClientSafeExceptionIfTypeIsMissing()
+    {
+        // Arrange
```
Comment (weisswurstkanone) [inline tests/GraphQL/Traits/ElementIdentificationTraitTest.php:48]: I am experiencing the following problem. Phpunit version is 7.5.20

1) ElementIdentificationTraitTest: Throwing client safe exception if type is missing
 Test  ../data-hub/tests/GraphQL/Traits/ElementIdentificationTraitTest.php:testThrowingClientSafeExceptionIfTypeIsMissing

  [Error] Call to undefined method Pimcore\Bundle\DataHubBundle\Tests\GraphQL\Traits\ElementIdentificationTraitTest::expectExceptionMessageMatches()

/home/pimcore-5/demo-basic/dev/pimcore/data-hub/tests/GraphQL/Traits/ElementIdentificationTraitTest.php:49
/home/pimcore-5/demo-basic/vendor/phpunit/phpunit/src/Framework/TestCase.php:1154
/home/pimcore-5/demo-basic/vendor/phpunit/phpunit/src/Framework/TestCase.php:842
/home/pimcore-5/demo-basic/vendor/phpunit/phpunit/src/Framework/TestResult.php:693

File: src/GraphQL/DataObjectType/PimcoreObjectType.php (hunk 7 above)
Comment (SimonBader) [reply src/GraphQL/DataObjectType/PimcoreObjectType.php]: I removed it again.

File: src/GraphQL/Resolver/QueryType.php (hunk 6 above)
Comment (SimonBader) [reply src/GraphQL/Resolver/QueryType.php:204]: Done.
PS: well, the @deprecated annotation was not easily applicable to an 'args' argument item.

[the remark this answers is not in the document]
Comment (SimonBader) [reply src/GraphQL/Resolver/QueryType.php:204]: Is there an easy way how to set up pimcore 5.7? I tried to do set up Pimcore skeleton with composer but I failed as the corresponding packages are whether not supported or not there anymore:
`docker run --rm --mount type=bind,source="$(pwd)",target=/var/www/html pimcore/pimcore:PHP7.2-apache /bin/bash -c "COMPOSER_MEMORY_LIMIT=-1 composer create-project -v pimcore/skeleton:v1.2.0 pimcore_5_7"`

File: src/GraphQL/Resolver/QueryType.php [hunk 9]
```diff
@@ -266,7 +269,8 @@ public function resolveObjectGetter($value = null, $args = [], $context, Resolve
         $objectList->setUnpublished(1);
         $objectList = $objectList->load();
         if (!$objectList) {
-            throw new ClientSafeException('object with ID ' . $args["id"] . ' not found');
+            $identifiers = ($args["id"] ? ' ID: ' . $args["id"] : '') . ($args["fullpath"] ? ' FULLPATH: ' . $args["fullpath"] : '');
+            throw new ClientSafeException('object with' . $identifiers . ' not found');
         }
         $object = $objectList[0];
```
Comment (SimonBader) [inline src/GraphQL/Resolver/QueryType.php:275]: @weisswurstkanone From what I tested so far line 237-275 can be replaced by one line of code calling the new trait:
`$object = $this->getElementByTypeAndIdOrPath($args, 'object');`

I am not sure about edge cases (e.g. setUnpublishd(1) on line 269) and I did not test it with Pimcore 5.7 yet.

File: src/GraphQL/Resolver/QueryType.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (weisswurstkanone) [reply src/GraphQL/Resolver/QueryType.php:204]: Could you please consult the forums regarding this question?

File: src/GraphQL/Resolver/QueryType.php (hunk 9 above)
Comment (weisswurstkanone) [reply src/GraphQL/Resolver/QueryType.php:273]: I don't think so because for objects there is this global sql conditional (a basic condition you can configure)
See https://github.com/pimcore/data-hub/blob/3ff7ada79b3d395911684e4627bec10af71f5bcb/src/GraphQL/Resolver/QueryType.php#L247

File: tests/GraphQL/Traits/ElementIdentificationTraitTest.php (hunk 8 above)
Comment (SimonBader) [reply tests/GraphQL/Traits/ElementIdentificationTraitTest.php:48]: Ah, finally I found it: the function expectExceptionMessageMatches() was initially named differently (renamed in phpunit 8.4): https://github.com/sebastianbergmann/phpunit/commit/d1199cb2e43a934b51521656be9748f63febe18e
For backward compatible (to phpunit 7.5 / Pimcore 5.7) I renamed the function back to expectExceptionMessageRegExp().

File: src/GraphQL/Resolver/QueryType.php (hunk 6 above)
Comment (SimonBader) [reply src/GraphQL/Resolver/QueryType.php:206]: I added a deprecation warning to the log (therefore I had to import the static Pimcore Logger to the Resolver/QueryType class though).

File: src/GraphQL/Resolver/QueryType.php (hunk 9 above)
Comment (SimonBader) [reply src/GraphQL/Resolver/QueryType.php:275]: Right, then I will leave it as is.

**Review by SimonBader — APPROVED** [review]

**Comment by weisswurstkanone** [comment]
@SimonBader the PR is still in draft state

**Comment by SimonBader** [comment]
> @SimonBader the PR is still in draft state

@weisswurstkanone @Iblis  I do not have the right to switch it to "Ready for review"

**Comment by Iblis** [comment]
@SimonBader I switched the PR to 'Ready for Review'

**Comment by weisswurstkanone** [comment]
I am getting a

{
  "errors": [
    {
      "message": "Notice: Undefined index: id"
    }
  ]
}

here 

https://github.com/pimcore/data-hub/blob/29cf29ab776652a8b0e2b9e72420b7313deac565/src/GraphQL/Resolver/QueryType.php#L272

Request was: 

```
{
  getCar(fullpath:"/") {
    color
  }
}
```


Could you please retest this with  error reporting set to E_ALL?

Just put a .env file in the project root with the following content:

```
PIMCORE_PHP_ERROR_REPORTING=32767
```

**Comment by SimonBader** [comment]
@weisswurstkanone I corrected the error handling.

**Comment by SimonBader** [comment]
> maybe just
> $isIdSet = $args["id"] ?? false;

I adapted it. For consistency reasons shall I change it for the [default language check](https://github.com/pimcore/data-hub/pull/256/files#diff-bc0b4125c17160c30d9a48cad9dffe9d769a3bcc135c61d3aeae249897a11aa5R237) as well?

**Comment by weisswurstkanone** [comment]
Yes please, would be nice!

**Comment by SimonBader** [comment]
> Yes please, would be nice!

Done

**Comment by weisswurstkanone** [comment]
Looks quite nice but I would not allow both ways in parallel.
This query will fail because it will try to match both ID and fullpath. Maybe catch this and throw a meaningful exception?

image

**Comment by SimonBader** [comment]
> Looks quite nice but I would not allow both ways in parallel.

Done

**Comment by weisswurstkanone** [comment]
Great. Thanks a lot!
